xbmc/cores/VideoPlayer/VideoPlayer.cpp

```diff
@@ -3134,9 +3141,48 @@ int CVideoPlayer::GetSubtitle()
 
 void CVideoPlayer::GetSubtitleStreamInfo(int index, SPlayerSubtitleStreamInfo &info)
 {
-  if (index < 0 || index > (int) GetSubtitleCount() - 1)
+  CSingleLock lock(m_player_status_lock);
+  if (index < 0 || index > (int) m_player_status_subtitle.size() - 1)
+    return;
+
+  info = m_player_status_subtitle[index];
+}
+
+void CVideoPlayer::GetVideoStreamInfo(int index, SPlayerVideoStreamInfo &info)
+{
+  CSingleLock lock(m_player_status_lock);
+  if (index < 0 || index > (int) m_player_status_video.size() - 1)
     return;
 
+  info = m_player_status_video[index];
+}
+
+void CVideoPlayer::GetAudioStreamInfo(int index, SPlayerAudioStreamInfo &info)
+{
+  CSingleLock lock(m_player_status_lock);
+  if (index < 0 || index > (int) m_player_status_audio.size() - 1)
+    return;
+
+  info = m_player_status_audio[index];
+}
+
+void CVideoPlayer::UpdateStreamInfos()
+{
+  CSingleLock lock(m_player_status_lock);
+  m_player_status_audio.resize(GetAudioStreamCount());
```

Comment:
The number of streams won't change unless player receives a stream change event. This event triggers OpenDefaultStreams. More than 90% of the info does not require to be updated periodically.

Why don't you add the parameters to SelectionStream and only update the variable parameters periodically?